### PR TITLE
[E2E] Epinio integration

### PR DESCRIPTION
**Pull request by evertonlperes**
# This PR is dead. See #1774

### Objective
Perform a full E2E testing, installing epinio on RD and pushing a sample app via epinio CLI.

### Changes
1. Create `epinio` spec file.
2. Created a sample php asset project (buildpack).
3. Set a fixed working epinio CLI version (`0.5.0`).

Note: This spec file will work only at local environment, for CI still disabled due some issues to handle traefik deployment (from k3s) through the tests. I'll work on a different card in order to fix the tests on CI (#1634)

### CODE REVIEW

File: e2e/epinio.e2e.spec.ts [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+/**
+ * Helper to identify the Load Balancer IP Address.
+ * It will return the traefik IP address, required by epinio install.
+ */
+export async function loadBalancerIp() {
+  const serviceInfo = await kubectl('describe', 'service', 'traefik', '--namespace', 'kube-system');
+
+  const serviceFiltered = serviceInfo.split('\n').toString();
+  const ipAddrRegex = /(LoadBalancer Ingress:)\s+(((?:[0-9]{1,3}\.){3}[0-9]{1,3}))/;
```
Comment (ericpromislow) [inline e2e/epinio.e2e.spec.ts:115]: I would simplify this as so:
```
  const m = /LoadBalancer Ingress:\s+(((?:[0-9]{1,3}\.){3}[0-9]{1,3}))/.exec(serviceFiltered);

  if (m) {
    return m[1];
  } else {
    console.log(`Cannot find load balancer IP address from ${ serviceFiltered }`);
  }
```

File: e2e/utils/TestUtils.ts [hunk 2]
```diff
@@ -2,7 +2,7 @@
  * TestUtils exports functions required for the E2E test specs.
  */
 import os from 'os';
-import fs from 'fs';
+import fs, { mkdirSync } from 'fs';
```
Comment (ericpromislow) [inline e2e/utils/TestUtils.ts:5]: Don't see a reason to single out `mkdirSync`

File: e2e/utils/TestUtils.ts [hunk 3]
```diff
@@ -37,13 +37,52 @@ export function playwrightReportAssets(fileName: string) {
...
+export function setUpHelmCustomEnv() {
+  const tempHelmFolder = path.join(os.homedir(), 'helmTmp');
+  const helmRepoCacheFolder = path.join(tempHelmFolder, 'Caches', 'helm', 'repository');
+  const helmRepoConfigFolder = path.join(tempHelmFolder, 'Config', 'helm', 'repository');
+
+  process.env.HELM_REPOSITORY_CACHE = helmRepoCacheFolder;
+  process.env.HELM_REPOSITORY_CONFIG = `${ helmRepoConfigFolder }/repositories.yaml`;
+
+  if (!fs.existsSync(tempHelmFolder)) {
+    mkdirSync(tempHelmFolder, { recursive: true });
```
Comment (ericpromislow) [inline e2e/utils/TestUtils.ts:53]: `fs.mkdirSync`

File: e2e/utils/TestUtils.ts [hunk 4]
```diff
@@ -37,13 +37,52 @@ export function playwrightReportAssets(fileName: string) {
...
+  const helmTempPath = path.join(os.homedir(), 'helmTmp');
+
+  if (fs.existsSync(helmTempPath)) {
+    fs.rmSync(helmTempPath, { recursive: true });
+  }
+}
+
+/**
+ * Detects which platform the spec is running and returns
+ * the platform name string.
```
Comment (ericpromislow) [inline e2e/utils/TestUtils.ts:71]: This seems unnecessary. Just check for an unrecognized platform through a hash object where `detectPlatform` is invoked.

**Review by ericpromislow — CHANGES_REQUESTED** [review]

File: e2e/epinio.e2e.spec.ts [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+  const regex = new RegExp(`${ ipAddrRegex.source }`);
+  const ipAddressLb = regex.exec(serviceFiltered);
+
+  // checking if it will be undefined, null, 0 or empty
+  if (typeof ipAddressLb !== 'undefined' && ipAddressLb) {
```
Comment (mook-as) [inline e2e/epinio.e2e.spec.ts:120]:
```suggestion
  if (ipAddressLb) {
```
If it's undefined, it's falsy.

File: e2e/epinio.e2e.spec.ts [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+/**
+ * Download epinio binary based on the platform type and save the binary
+ * into a temporary folder.
+ */
+export async function downloadEpinioBinary( platformType: string) {
+  // Setting up epinio binaries names per platform
+  const epinioWin = 'epinio-windows-amd64.exe';
+  const epinioLinux = 'epinio-linux-x86_64';
+  const epinioDarwin = 'epinio-darwin-x86_64';
+  const epinioDarwinArm = 'epinio-darwin-arm64';
```
Comment (mook-as) [inline e2e/epinio.e2e.spec.ts:144]: This should be a map keyed by platform / arch.

File: e2e/epinio.e2e.spec.ts [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+  const epinioWorkingVersion = 'v0.5.0';
+
+  // Create a temp folder for epinio binary
+  const epinioTempFolder = path.join(os.homedir(), 'epinio-tmp');
```
Comment (mook-as) [inline e2e/epinio.e2e.spec.ts:148]: Should probably use [`os.tmpdir()`](https://nodejs.org/docs/latest-v16.x/api/os.html#ostmpdir).

File: e2e/epinio.e2e.spec.ts [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+ * Gracefully remove epinio temp folder and uninstall all epinio-install resources
+ */
+export async function tearDownEpinio() {
+  const epinioTempFolder = path.join(os.homedir(), 'epinio-tmp');
+
+  if (fs.existsSync(epinioTempFolder)) {
+    fs.rmSync(epinioTempFolder, { recursive: true, maxRetries: 10 });
+  }
+
+  await helm('uninstall', 'epinio-installer', '--timeout=20m');
```
Comment (mook-as) [inline e2e/epinio.e2e.spec.ts:204]: Doesn't this need `--wait`, given that we're about to remove helm?

File: e2e/epinio.e2e.spec.ts [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+/**
+ * Run the given tool with the given arguments, returning its standard output.
+ */
+export async function tool(tool: string, ...args: string[]): Promise<string> {
```
Comment (mook-as) [inline e2e/epinio.e2e.spec.ts:210]: Why are we not using the `TestUtils.ts` version?

**Review by mook-as — CHANGES_REQUESTED** [review]

File: e2e/epinio.e2e.spec.ts (hunk 5 above)
Comment (ericpromislow) [reply e2e/epinio.e2e.spec.ts:120]: Done.

File: e2e/utils/TestUtils.ts (hunk 2 above)
Comment (ericpromislow) [reply e2e/utils/TestUtils.ts:5]: Done.

File: e2e/utils/TestUtils.ts (hunk 3 above)
Comment (ericpromislow) [reply e2e/utils/TestUtils.ts:53]: done

File: e2e/utils/TestUtils.ts [hunk 10]
```diff
@@ -37,13 +37,52 @@ export function playwrightReportAssets(fileName: string) {
...
+ */
+export function detectPlatform() {
```
[the remark this answers is not in the document]
Comment (ericpromislow) [reply e2e/utils/TestUtils.ts:73]: Yeah, I removed it as well.

**Comment by mook-as** [comment]
Closing PR in favour of #1774.

**Pull request closed without merging**